tool/tsh/db.go

```diff
@@ -41,19 +42,63 @@ func onListDatabases(cf *CLIConf) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
 	var databases []types.Database
 	err = client.RetryWithRelogin(cf.Context, tc, func() error {
 		databases, err = tc.ListDatabases(cf.Context)
 		return trace.Wrap(err)
 	})
+
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
 	// Retrieve profile to be able to show which databases user is logged into.
 	profile, err := client.StatusCurrent(cf.HomePath, cf.Proxy)
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
+	// calculating possible users for db is only done in verbose mode.
+	displayUsersForDb := func(database types.Database) string {
+		return ""
+	}
+
+	if cf.Verbose {
+		// in case of any errors down the line, indicate that fact.
+		displayUsersForDb = func(database types.Database) string {
+			return "(error: unable to check)"
+		}
+
+		err = client.RetryWithRelogin(cf.Context, tc, func() error {
+			proxy, err := tc.ConnectToProxy(cf.Context)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			cluster, err := proxy.ConnectToCluster(cf.Context, tc.SiteName, true)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			defer cluster.Close()
+
+			roleSet, err := services.FetchRoles(profile.Roles, cluster, profile.Traits)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			displayUsersForDb = func(database types.Database) string {
+				return getUsersForDb(roleSet, database)
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			log.Debugf("Error while calculating available db users: %v", err)
```

Comment:
Not sure if the flow should processed when a user is unable to connect to the proxy

Reply:
Potentially this can also be caused by a lack of permissions. But I'm pretty sure this part will change by moving the calculation logic to the server and exposing it via API.